File: auto_bot_compund.js
const fs = require('fs');
const axios = require('axios');
const PDFDocument = require('pdfkit');
const { DateTime } = require('luxon');
const snoowrap = require('snoowrap');

// === CONFIG ===
const CAPITAL_FILE = 'capital.json';
const TRADE_LOG_FILE = 'trades_history.json';
const SIGNAL_DIR = 'signals';
const TRADE_DIR = 'trades';
const START_CAPITAL = 10.0;
const MAX_LOSS_PCT = 15;
const TP_PERCENT = 0.25;
const SL_PERCENT = 0.10;
const LEVERAGE = 20;

const DISCORD_WEBHOOK_URL = 'https://discord.com/api/webhooks/...';
const REDDIT_CREDS = {
  userAgent: 'cryptopilot_bot',
  clientId: 'YOUR_CLIENT_ID',
  clientSecret: 'YOUR_CLIENT_SECRET',
  username: 'YOUR_USERNAME',
  password: 'YOUR_PASSWORD',
};

if (!fs.existsSync(SIGNAL_DIR)) fs.mkdirSync(SIGNAL_DIR);
if (!fs.existsSync(TRADE_DIR)) fs.mkdirSync(TRADE_DIR);

// === DISCORD + REDDIT ===
function postToDiscord(message) {
  axios.post(DISCORD_WEBHOOK_URL, { content: message }).catch(e => console.log('❌ Discord Error:', e.message));
}

function postToReddit(title, body) {
  try {
    const reddit = new snoowrap(REDDIT_CREDS);
    reddit.getSubreddit('YourSubreddit').submitSelfpost({ title, text: body });
  } catch (e) {
    console.log('❌ Reddit Error:', e.message);
  }
}

// === TECHNICAL INDICATORS ===
function ema(values, period) {
  let k = 2 / (period + 1);
  let emaArr = [], emaPrev = values.slice(0, period).reduce((a, b) => a + b) / period;
  emaArr.push(emaPrev);
  for (let i = period; i < values.length; i++) {
    emaPrev = values[i] * k + emaPrev * (1 - k);
    emaArr.push(emaPrev);
  }
  return Array(period - 1).fill(null).concat(emaArr);
}

function sma(values, period) {
  return values.map((_, i) =>
    i < period - 1 ? null : values.slice(i + 1 - period, i + 1).reduce((a, b) => a + b) / period
  );
}

function computeRSI(closes, period = 14) {
  let gains = [], losses = [];
  for (let i = 1; i < closes.length; i++) {
    const delta = closes[i] - closes[i - 1];
    gains.push(Math.max(delta, 0));
    losses.push(Math.max(-delta, 0));
  }
  if (gains.length < period) return 50;
  const avgGain = gains.slice(-period).reduce((a, b) => a + b) / period;
  const avgLoss = losses.slice(-period).reduce((a, b) => a + b) / period;
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return Math.round(100 - (100 / (1 + rs)) * 100) / 100;
}

function calculateMACD(values, fast = 12, slow = 26, signal = 9) {
  const emaFast = ema(values, fast);
  const emaSlow = ema(values, slow);
  const macd = emaFast.map((v, i) => v !== null && emaSlow[i] !== null ? v - emaSlow[i] : null);
  const macdFiltered = macd.filter(x => x !== null);
  const signalLine = ema(macdFiltered, signal);
  const paddedSignal = Array(macd.length - signalLine.length).fill(null).concat(signalLine);
  const histogram = macd.map((v, i) => (v !== null && paddedSignal[i] !== null ? v - paddedSignal[i] : null));
  return { macd, signal: paddedSignal, histogram };
}

function calculateBollingerBands(values, period = 20, stdDev = 2) {
  const smaVals = sma(values, period);
  return values.map((_, i) => {
    if (i < period - 1) return [null, null, null];
    const mean = smaVals[i];
    const std = Math.sqrt(values.slice(i + 1 - period, i + 1).reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / period);
    return [mean + stdDev * std, mean, mean - stdDev * std];
  });
}

// === HELPERS ===
function loadCapital() {
  if (!fs.existsSync(CAPITAL_FILE)) return START_CAPITAL;
  return JSON.parse(fs.readFileSync(CAPITAL_FILE)).balance || START_CAPITAL;
}

function saveCapital(balance) {
  fs.writeFileSync(CAPITAL_FILE, JSON.stringify({ balance: parseFloat(balance.toFixed(4)) }));
}

function logTrade(trade) {
  let trades = fs.existsSync(TRADE_LOG_FILE) ? JSON.parse(fs.readFileSync(TRADE_LOG_FILE)) : [];
  trades.push(trade);
  fs.writeFileSync(TRADE_LOG_FILE, JSON.stringify(trades, null, 2));
}

function todayLossPct() {
  if (!fs.existsSync(TRADE_LOG_FILE)) return 0;
  const trades = JSON.parse(fs.readFileSync(TRADE_LOG_FILE));
  const today = DateTime.utc().toISODate();
  const loss = trades.filter(t => t.timestamp.startsWith(today) && t.pnl < 0).reduce((sum, t) => sum + t.pnl, 0);
  const capital = loadCapital();
  return capital > 0 ? Math.round((-loss / capital) * 10000) / 100 : 100;
}

function formatSignal(s) {
  return `📈 ${s.symbol} [${s.side}] | ${s.strategy}
Entry: ${s.entry} | TP: ${s.tp} | SL: ${s.sl}
Confidence: ${s.confidence}% | Score: ${s.score}
Regime: ${s.regime} | Trend: ${s.trend}
Timestamp: ${s.timestamp}`;
}

function saveJSON(data, path) {
  fs.writeFileSync(path, JSON.stringify(data, null, 2));
}

function savePDF(filename, items, title) {
  const doc = new PDFDocument();
  doc.pipe(fs.createWriteStream(filename));
  doc.fontSize(14).text(title, { align: 'center' }).moveDown();
  items.forEach(item => {
    Object.entries(item).forEach(([k, v]) => {
      doc.fontSize(10).text(`${k}: ${v}`);
    });
    doc.moveDown();
  });
  doc.end();
}

// === DEMO TRADE ===
function simulateTrade(signal) {
  const entry = signal.entry, tp = signal.tp;
  const side = signal.side;
  let capital = loadCapital();
  const riskAmount = capital * 0.02;
  const riskPerUnit = Math.abs(entry - signal.sl);
  const qty = riskPerUnit > 0 ? parseFloat((riskAmount / riskPerUnit).toFixed(4)) : 0;
  const pnl = parseFloat(((side === 'LONG' ? tp - entry : entry - tp) * qty).toFixed(4));
  capital += pnl;
  saveCapital(capital);
  const trade = {
    symbol: signal.symbol,
    side,
    entry,
    exit: tp,
    qty,
    pnl,
    strategy: signal.strategy,
    timestamp: DateTime.utc().toISO({ suppressMilliseconds: true }),
  };
  logTrade(trade);
  return trade;
}
async function main() {
  console.log('🚀 Running CryptoPilot...');
  const balance = loadCapital();
  const lossPct = todayLossPct();

  if (lossPct >= MAX_LOSS_PCT) {
    console.log(`❌ Max daily loss of ${MAX_LOSS_PCT}% hit (${lossPct}%) — Trading paused.`);
    return;
  }

  const allSignals = [];
  const trades = [];

  const symbols = await getSymbols(30);
  for (const symbol of symbols) {
    const signals = await analyze(symbol);
    allSignals.push(...signals);
  }

  if (allSignals.length === 0) {
    console.log('⚠️ No signals found.');
    return;
  }

  const top5 = allSignals
    .sort((a, b) => b.score - a.score || b.confidence - a.confidence)
    .slice(0, 5);

  for (let i = 0; i < top5.length; i++) {
    const s = top5[i];
    console.log(`#${i + 1} ${s.symbol} ${s.side} | Score: ${s.score} | ${s.strategy}`);
    const msg = formatSignal(s);
    saveJSON(s, `${SIGNAL_DIR}/${s.symbol}.json`);
    postToDiscord(msg);
    postToReddit(`CryptoPilot Signal - ${s.symbol} #${i + 1}`, msg);

    const trade = simulateTrade(s);
    saveJSON(trade, `${TRADE_DIR}/${s.symbol}.json`);
    trades.push(trade);

    postToDiscord(`🟢 Executed Trade: ${trade.symbol} | PnL: ${trade.pnl}`);
    postToReddit(`CryptoPilot Trade - ${trade.symbol}`, `Trade Summary:\n\`\`\`\n${JSON.stringify(trade, null, 2)}\n\`\`\``);
  }

  savePDF('all_signals.pdf', allSignals, '📊 All Signals');
  savePDF('all_trades.pdf', trades, '💹 Executed Trades');

  console.log('\n✅ Run complete. Trades and signals exported.\n');
}

main();

File: auto_bot_compund.py
import os, json, requests
from datetime import datetime, timedelta, timezone
from fpdf import FPDF
import praw # type: ignore

# === CONFIG ===
CAPITAL_FILE = "capital.json"
TRADE_LOG_FILE = "trades_history.json"
SIGNAL_DIR = "signals"
TRADE_DIR = "trades"
START_CAPITAL = 10.0
MAX_LOSS_PCT = 15
TP_PERCENT = 0.25
SL_PERCENT = 0.10
LEVERAGE = 20

DISCORD_WEBHOOK_URL = "https://discord.com/api/webhooks/..."  # Replace with your real webhook
REDDIT_CREDS = {
    "client_id": "YOUR_CLIENT_ID",
    "client_secret": "YOUR_CLIENT_SECRET",
    "username": "YOUR_USERNAME",
    "password": "YOUR_PASSWORD",
    "user_agent": "cryptopilot_bot"
}

for d in [SIGNAL_DIR, TRADE_DIR]:
    os.makedirs(d, exist_ok=True)

# === REDDIT + DISCORD ===
def post_to_discord(message):
    try:
        requests.post(DISCORD_WEBHOOK_URL, json={"content": message})
    except Exception as e:
        print("❌ Discord Error:", e)

def post_to_reddit(title, body):
    try:
        reddit = praw.Reddit(**REDDIT_CREDS)
        subreddit = reddit.subreddit("YourSubreddit")
        subreddit.submit(title, selftext=body)
    except Exception as e:
        print("❌ Reddit Error:", e)

# === INDICATORS ===
def ema(values, period):
    emas, k = [], 2 / (period + 1)
    ema_prev = sum(values[:period]) / period
    emas.append(ema_prev)
    for price in values[period:]:
        ema_prev = price * k + ema_prev * (1 - k)
        emas.append(ema_prev)
    return [None] * (period - 1) + emas

def sma(values, period):
    return [None if i < period - 1 else sum(values[i+1-period:i+1]) / period for i in range(len(values))]

def compute_rsi(closes, period=14):
    gains, losses = [], []
    for i in range(1, len(closes)):
        delta = closes[i] - closes[i - 1]
        gains.append(max(delta, 0))
        losses.append(max(-delta, 0))
    if len(gains) < period: return 50
    avg_gain = sum(gains[-period:]) / period
    avg_loss = sum(losses[-period:]) / period
    if avg_loss == 0: return 100
    rs = avg_gain / avg_loss
    return round(100 - (100 / (1 + rs)), 2)

def calculate_macd(values, fast=12, slow=26, signal=9):
    ema_fast = ema(values, fast)
    ema_slow = ema(values, slow)
    macd_line = [f - s if f and s else None for f, s in zip(ema_fast, ema_slow)]
    signal_line = ema([x for x in macd_line if x is not None], signal)
    signal_line = [None] * (len(macd_line) - len(signal_line)) + signal_line
    histogram = [m - s if m and s else None for m, s in zip(macd_line, signal_line)]
    return macd_line, signal_line, histogram

def calculate_bollinger_bands(values, period=20, std_dev=2):
    sma_vals = sma(values, period)
    bands = []
    for i in range(len(values)):
        if i < period - 1:
            bands.append((None, None, None))
        else:
            mean = sma_vals[i]
            std = (sum((x - mean) ** 2 for x in values[i + 1 - period:i + 1]) / period) ** 0.5
            upper = mean + std_dev * std
            lower = mean - std_dev * std
            bands.append((upper, mean, lower))
    return bands

def detect_market_trend(symbol):
    def fetch_closes(symbol, tf):
        url = f"https://fapi.binance.com/fapi/v1/klines?symbol={symbol}&interval={tf}&limit=60"
        try:
            r = requests.get(url, timeout=5)
            return [float(x[4]) for x in r.json()]
        except:
            return []
    trend_info = {}
    for tf in ['1h', '4h', '15m']:
        closes = fetch_closes(symbol, tf)
        if len(closes) < 50:
            trend_info[tf] = 'neutral'
            continue
        ema9 = ema(closes, 9)[-1]
        ema21 = ema(closes, 21)[-1]
        ma200 = sma(closes, 50)[-1]
        close = closes[-1]
        if close > ma200 and ema9 > ema21:
            trend_info[tf] = 'bullish'
        elif close < ma200 and ema9 < ema21:
            trend_info[tf] = 'bearish'
        else:
            trend_info[tf] = 'neutral'
    return trend_info
# === HELPER: CAPITAL TRACKING ===
def load_capital():
    if not os.path.exists(CAPITAL_FILE):
        return START_CAPITAL
    with open(CAPITAL_FILE) as f:
        return json.load(f).get("balance", START_CAPITAL)

def save_capital(balance):
    with open(CAPITAL_FILE, "w") as f:
        json.dump({"balance": round(balance, 4)}, f)

# === HELPER: TRADE HISTORY ===
def log_trade(trade):
    trades = []
    if os.path.exists(TRADE_LOG_FILE):
        with open(TRADE_LOG_FILE) as f:
            trades = json.load(f)
    trades.append(trade)
    with open(TRADE_LOG_FILE, "w") as f:
        json.dump(trades, f, indent=2)

def today_loss_pct():
    if not os.path.exists(TRADE_LOG_FILE): return 0
    with open(TRADE_LOG_FILE) as f:
        trades = json.load(f)
    today = datetime.utcnow().strftime("%Y-%m-%d")
    loss = sum(t["pnl"] for t in trades if t["timestamp"].startswith(today) and t["pnl"] < 0)
    capital = load_capital()
    return round(-loss / capital * 100, 2) if capital > 0 else 100

# === SIGNAL FORMATTING + EXPORT ===
def format_signal(s):
    return f"""📈 {s['symbol']} [{s['side']}] | {s['strategy']}
Entry: {s['entry']} | TP: {s['tp']} | SL: {s['sl']}
Confidence: {s['confidence']}% | Score: {s['score']}
Regime: {s['regime']} | Trend: {s['trend']}
Timestamp: {s['timestamp']}"""

def save_json(data, path):
    with open(path, "w") as f:
        json.dump(data, f, indent=2)

def save_pdf(filename, items, title):
    pdf = FPDF()
    pdf.add_page()
    pdf.set_font("Arial", size=12)
    pdf.cell(0, 10, title, ln=True, align="C")
    pdf.ln(5)
    for item in items:
        for k, v in item.items():
            pdf.multi_cell(0, 8, f"{k}: {v}")
        pdf.ln(5)
    pdf.output(filename)

# === DEMO TRADE + COMPOUND ===
def simulate_trade(signal):
    entry, tp = signal['entry'], signal['tp']
    side = signal['side']
    capital = load_capital()
    risk_amount = capital * 0.02
    risk_per_unit = abs(signal['entry'] - signal['sl'])
    qty = round(risk_amount / risk_per_unit, 4) if risk_per_unit > 0 else 0
    pnl = round((tp - entry) * qty if side == "LONG" else (entry - tp) * qty, 4)
    capital += pnl
    save_capital(capital)
    trade = {
        "symbol": signal["symbol"],
        "side": side,
        "entry": entry,
        "exit": tp,
        "qty": qty,
        "pnl": pnl,
        "strategy": signal["strategy"],
        "timestamp": signal["timestamp"]
    }
    log_trade(trade)
    return trade
def fetch_ohlcv(symbol, interval='1h', limit=100):
    url = f"https://fapi.binance.com/fapi/v1/klines?symbol={symbol}&interval={interval}&limit={limit}"
    try:
        r = requests.get(url, timeout=5)
        return [[float(x[2]), float(x[3]), float(x[4]), float(x[5]), float(x[1])] for x in r.json()]
    except:
        return []

def get_symbols(limit=100):
    try:
        r = requests.get("https://fapi.binance.com/fapi/v1/exchangeInfo", timeout=5)
        return [s['symbol'] for s in r.json()['symbols'] if s['contractType'] == 'PERPETUAL' and 'USDT' in s['symbol']][:limit]
    except:
        return []

def build_signal(name, condition, confidence, regime, trend_info, close, symbol, tf, rsi, macd_hist, bb_upper, bb_lower, volumes):
    if not condition:
        return None
    side = "long" if name != "Short Reversal" else "short"
    if not is_trade_allowed(side.upper(), trend_info): return None
    entry = close
    sl_price = entry * (1 - SL_PERCENT) if side == "long" else entry * (1 + SL_PERCENT)
    tp_price = entry * (1 + TP_PERCENT) if side == "long" else entry * (1 - TP_PERCENT)
    liquidation = entry * (1 - 1 / LEVERAGE) if side == "long" else entry * (1 + 1 / LEVERAGE)
    signal = {
        "symbol": symbol,
        "timeframe": tf,
        "side": side.upper(),
        "entry": round(entry, 6),
        "sl": round(sl_price, 6),
        "tp": round(tp_price, 6),
        "liquidation": round(liquidation, 6),
        "rsi": rsi,
        "macd_hist": round(macd_hist[-1], 4) if macd_hist[-1] else None,
        "bb_breakout": "YES" if close > bb_upper[-1] or close < bb_lower[-1] else "NO",
        "trend": "bullish" if side == "long" else "bearish",
        "regime": regime,
        "confidence": confidence,
        "score": 80 + confidence * 0.2,
        "strategy": name,
        "timestamp": (datetime.now(timezone.utc) + timedelta(hours=3)).strftime("%Y-%m-%d %H:%M UTC+3"),
        "vol_spike": volumes[-1] > sum(volumes[-20:]) / 20 * 1.5
    }
    return signal

def is_trade_allowed(side, trend_info):
    trend_votes = list(trend_info.values())
    bull = trend_votes.count('bullish')
    bear = trend_votes.count('bearish')
    if bull > bear and side == 'SHORT': return False
    if bear > bull and side == 'LONG': return False
    return True

def analyze(symbol, tf="1h"):
    data = fetch_ohlcv(symbol, tf)
    if len(data) < 60: return []
    highs = [x[0] for x in data]
    lows = [x[1] for x in data]
    closes = [x[2] for x in data]
    volumes = [x[3] for x in data]
    close = closes[-1]
    ema9 = ema(closes, 9)
    ema21 = ema(closes, 21)
    ma20 = sma(closes, 20)
    ma200 = sma(closes, 50)
    rsi = compute_rsi(closes)
    bb_upper, bb_mid, bb_lower = zip(*calculate_bollinger_bands(closes))
    macd_line, macd_signal, macd_hist = calculate_macd(closes)
    trend_info = detect_market_trend(symbol)
    regime = "trend" if ma20[-1] > ma200[-1] else "mean_reversion" if rsi < 35 or rsi > 65 else "scalp"
    signals = []
    if regime == "trend":
        sig = build_signal("Trend", ema9[-1] > ema21[-1], 90, regime, trend_info, close, symbol, tf, rsi, macd_hist, bb_upper, bb_lower, volumes)
        if sig: signals.append(sig)
    if regime == "mean_reversion":
        sig = build_signal("Mean-Reversion", rsi < 40 or close < ma20[-1], 85, regime, trend_info, close, symbol, tf, rsi, macd_hist, bb_upper, bb_lower, volumes)
        if sig: signals.append(sig)
    if regime == "scalp" and volumes[-1] > sum(volumes[-20:]) / 20 * 1.5:
        sig = build_signal("Scalp Breakout", True, 80, regime, trend_info, close, symbol, tf, rsi, macd_hist, bb_upper, bb_lower, volumes)
        if sig: signals.append(sig)
    if rsi > 65 and close > bb_upper[-1]:
        sig = build_signal("Short Reversal", True, 75, "reversal", trend_info, close, symbol, tf, rsi, macd_hist, bb_upper, bb_lower, volumes)
        if sig: signals.append(sig)
    return signals
def main():
    print("🚀 Running CryptoPilot...")
    balance = load_capital()
    loss_pct = today_loss_pct()

    if loss_pct >= MAX_LOSS_PCT:
        print(f"❌ Max daily loss of {MAX_LOSS_PCT}% hit ({loss_pct}%) — Trading paused.")
        return

    all_signals, trades = [], []

    for symbol in get_symbols(limit=30):
        signals = analyze(symbol)
        all_signals.extend(signals)

    if not all_signals:
        print("⚠️ No signals found.")
        return

    top5 = sorted(all_signals, key=lambda x: (x['score'], x['confidence']), reverse=True)[:5]

    # === SAVE + POST EACH SIGNAL ===
    for i, s in enumerate(top5, 1):
        print(f"\n#{i} {s['symbol']} {s['side']} | Score: {s['score']} | {s['strategy']}")
        msg = format_signal(s)
        save_json(s, f"{SIGNAL_DIR}/{s['symbol']}.json")
        post_to_discord(msg)
        post_to_reddit(f"CryptoPilot Signal - {s['symbol']} #{i}", msg)

        # === DEMO TRADE
        trade = simulate_trade(s)
        save_json(trade, f"{TRADE_DIR}/{s['symbol']}.json")
        trades.append(trade)
        post_to_discord(f"🟢 Executed Trade: {trade['symbol']} | PnL: {trade['pnl']}")
        post_to_reddit(f"CryptoPilot Trade - {trade['symbol']}", f"Trade Summary:\n{json.dumps(trade, indent=2)}")

    # === EXPORT PDFS
    save_pdf("all_signals.pdf", all_signals, "📊 All Signals")
    save_pdf("all_trades.pdf", trades, "💹 Executed Trades")

    print("\n✅ Run complete. Trades and signals exported.\n")

if __name__ == "__main__":
    main()
